src/aspire/orientation/__init__.py

```diff
@@ -1,38 +1,503 @@
+            is of size 2N x N(N_check-1)/2 (2N unknowns and N(N_check-1)/2 equations).
+            This may be too big if N is large. The algorithm will use `equations_factor`
+            times the total number of equations if the resulting total number of
+            memory requirements is less than `max_memory` (in megabytes); otherwise it
+            will reduce the number of equations to fit in `max_memory`.
+        :return: Estimated number of shift equations
         """
-        perform estimation of orientations
+        # Number of equations that will be used to estimation the shifts
+        n_equations_total = int(np.ceil(n_img * (self.n_check - 1) / 2))
+        # Estimated memory requirements for the full system of equation.
+        # This ignores the sparsity of the system, since backslash seems to
+        # ignore it.
+        memory_total = equations_factor * (
+                n_equations_total * 2 * n_img * np.dtype('float64').itemsize)
+        if memory_total < (max_memory * 10 ** 6):
+            n_equations = n_equations_total
+        else:
+            subsampling_factor = (max_memory * 10 ** 6) / memory_total
+            subsampling_factor = min(1.0, subsampling_factor)
+            n_equations = int(np.ceil(n_equations_total * subsampling_factor))
+
+        if n_equations < n_img:
+            logger.warning('Too few equations. Increase memory_factor. Setting n_equations to n_img.')
+            n_equations = n_img
+
+        if n_equations < 2 * n_img:
+            logger.warning('Number of equations is small. Consider increase memory_factor.')
+
+        return n_equations
+
+    def _generate_shift_phase_and_filter(self, r_max, max_shift, shift_step):
+        """
+        Prepare the shift phases and generate filter for common-line detection
+
+        The shift phases are pre-defined in a range of max_shift that can be
+        applied to maximize the common line calculation. The common-line filter
+        is also applied to the radial direction for easier detection.
+        :param r_max: Maximum index for common line detection
+        :param max_shift: Maximum value of 1D shift (in pixels) to search
+        :param shift_step: Resolution of shift estimation in pixels
+        :return: shift phases matrix and common lines filter
         """
-        pass
 
-    def output(self):
+        # Number of shifts to try
+        n_shifts = int(np.ceil(2 * max_shift / shift_step + 1))
+
+        # only half of ray
+        rk = np.arange(-r_max, 0)
+
+        # Generate all shift phases
+        shifts = -max_shift + shift_step * np.arange(n_shifts)
+        shift_phases = np.exp(
+            np.outer(-2 * np.pi * 1j * rk / (2 * r_max + 1), shifts))
+        # Set filter for common-line detection
+        h = np.sqrt(np.abs(rk)) * np.exp(-np.square(rk) / (2 * (r_max / 4) ** 2))
+        return shifts, shift_phases, h
+
+    def _generate_index_pairs(self, n_equations):
```

Comment:
note in another review comment I mention a method in numpy which yields exactly triu/tril indices directly. I pushed up a small PR #210 with a better implementation we can re-use.

Reply:
Thanks. I am not sure we need to use this or not. Is this will change the resulted random sequence? Then we need to rebuild unit tests.

Reply:
Sure. We've discussed several times in our weekly meeting the unit tests you are referring to being of marginal quality, ideas to improve them, including that they should to be rewritten (despite being new code).  We punted on that for now. I guess we should punt on this too for the same reason. However, that other PR 210 should go in, one way or another, and an issue about replacing this code created. It can be dealt with at the time the unit tests are.  Does that sound reasonable? Thanks

Reply:
Yes. I agree.